eng/psakefile.ps1

```diff
@@ -114,165 +135,132 @@ Task docs -Depends wheelhouse {
-            Write-BuildLog "LLVM target $($llvmTag) is already installed."
+            Write-BuildLog "LLVM target is already installed."
             # LLVM is already downloaded
             Use-LlvmInstallation $packagePath
         }
         else {
-            Write-BuildLog "LLVM target $($llvmTag) is not installed."
+            Write-BuildLog "LLVM target is not installed."
             if (Test-AllowedToDownloadLlvm) {
-                Write-BuildLog "Downloading LLVM target $packageName "
-                Install-LlvmFromBuildArtifacts $packagePath
+                Write-BuildLog "Downloading LLVM target"
+                Invoke-Task "install-llvm-from-archive"
             }
             else {
                 Write-BuildLog "Downloading LLVM Disabled, building from source."
                 # We don't have an external LLVM installation specified
                 # We are not downloading LLVM
                 # So we need to build it.
-                Install-LlvmFromSource $packagePath
+                Invoke-Task "install-llvm-from-source"
             }
         }
     }
 }
 
+task install-llvm-from-archive {
+    $cache = Resolve-InstallationDirectory
+    Use-LlvmInstallation $cache
+    $clear_cache_var = $false
+    if (!(Test-Path env:\QIRLIB_CACHE_DIR)) {
+        $clear_cache_var = $true
+        $env:QIRLIB_CACHE_DIR = $cache
+    }
+    try {
+        Invoke-LoggedCommand -wd $pyqir.qirlib.dir {
+            cargo build --release --no-default-features --features "download-llvm,no-llvm-linking" -vv
+        }
+    }
+    finally {
+        if ($clear_cache_var) {
+            Remove-Item -Path Env:QIRLIB_CACHE_DIR
+        }
+    }
+}
+
+task install-llvm-from-source {
+    $installationDirectory = Resolve-InstallationDirectory
+    Use-LlvmInstallation $cache
+    if (Test-RunInContainer) {
+        Build-ContainerImage $repo.root
+        $srcPath = $repo.root
+        $ioVolume = "$($srcPath):/io"
+        $llvmVolume = "$($installationDirectory):/usr/lib/llvm"
+        $userSpec = ""
+        Invoke-LoggedCommand {
+            docker run --rm $userSpec -v $ioVolume -v $llvmVolume -e QIRLIB_CACHE_DIR="/usr/lib/llvm" -w /io/qirlib manylinux2014_x86_64_maturin conda run --no-capture-output cargo build --release --no-default-features --features "build-llvm,no-llvm-linking" -vv
+        }
+    }
+    else {
+        if ($IsWindows) {
+            Include vcvars.ps1
+        }
+        $clear_cache_var = $false
+        if (!(Test-Path env:\QIRLIB_CACHE_DIR)) {
+            $clear_cache_var = $true
+            $env:QIRLIB_CACHE_DIR = $cache
+        }
+        try {
+            Invoke-LoggedCommand -wd $pyqir.qirlib.dir {
+                cargo build --release --no-default-features --features "build-llvm,no-llvm-linking" -vv
+            }
+        }
+        finally {
+            if ($clear_cache_var) {
+                Remove-Item -Path Env:QIRLIB_CACHE_DIR
+            }
+        }
```

Comment:
The lines 216 - 230 look nearly identical to lines 182 - 196. (the only diff I see is in line 223 - "download" vs. "build")
Consider extracting into a separate "function".

Reply:
247-261 is a different variable. I'll take a look.

Reply:
Pushed an update